ministryofjustice/analytics-platform-ops
# API Gateway resources to access to POST /github endpoint

## What

Terraform resources to deploy the API Gateway to handle the GitHub webhooks.


## Why

To be able to deploy the API very quickly without having to click around in the AWS Console (which is also error prone)


## Ticket

https://trello.com/c/YkoHJIMD/113-create-aws-api-gateway-with-terraform


## How to review

1. Delete the `dev_analytics` API [from the AWS Console](https://eu-west-1.console.aws.amazon.com/apigateway/home?region=eu-west-1#/apis)
2. Run `$ terraform plan/apply -target=module.data_access` from within `/infra/terraform/environments/dev`
3. Go to the [GitHub webhooks page](https://github.com/organizations/moj-analytical-services/settings/hooks/13614174#delivery-response), re-deliver one of the last events 
4. You should get back a `202 ACCEPTED` or a `404 NOT FOUND`, not a `500 INTERNAL SERVER ERROR` 


## About environment isolation

We decided to have an API Gateway per environment (called
`${ENV}_analytics`) with a stage called after the environment
(e.g. `dev`, `alpha`, etc...).
The stage will became part of the API URL, e.g. [https://abc123.execute-api.eu-west-1.amazonaws.com/dev](https://abc123.execute-api.eu-west-1.amazonaws.com/dev)
which is very unsurprising as you can always see the environment you're
calling.

## POST /github endpoint

At the moment the API is composed only by the following endpoint:

 - `POST /github`
   - Publishes the GitHub event to SNS and returns `202 ACCEPTED`
   - If the SNS doesn't exist (because we're not interested in handle it),
   it returnd `404 NOT FOUND`
   - If the secret provided by GitHub (via `X-Hub-Signature` header) is
   invalid returns `401 UNAUTHORIZED`


## Terraform details

We decided to have 3 modules to avoid dump everything in a single module:
- `webhooks_api` contains the API Gateway resources
- `notifications` contains the SNS topics and the lambda function
   which publishes the GitHub events to these topics
- `data_access` will contain (currently empty) the lambda functions
   to handle these events and add/delete IAM resources

In AWS API GateWay parlance these are the terraform resources created:
 - `*_gateway_rest_api` for the whole API
 - `*_deployment` the deployment which makes the API methods available from
 the outside. APIs are deployed to a "stage"
 - `*_resource` is the resource which can manipulated via its methods,
 `/github` in our case
 - `*_method` is the method to manipulate a specific resource,
it's the combination of a HTTP verb and the resource name, e.g. `POST /github`.
Note this method **does not requires** authentication because the Lambda
function will check the GitHub webhook signature
 - `*_integration` is the bit which converts the request format
 to the "backend" endpoint's format. In our case we just use the Lambda Proxy
 integration, which means the request goes straight to the Lambda function
 without alterations
 - `*_role`, the IAM role assumed by the API when calling the Lambda function
 - `*_role_policy`, the IAM policy which grants the role the permission to
 invoke the Lambda function
